File: liv/src/main/java/com/benitobertoli/liv/rule/PatternRule.java
package com.benitobertoli.liv.rule;

import android.support.annotation.NonNull;

import java.util.regex.Pattern;

import io.reactivex.Observable;

public class PatternRule extends BaseRule {

    protected Pattern pattern;

    public PatternRule(String errorMessage, Pattern pattern) {
        super(errorMessage);
        if (pattern == null) {
            throw new IllegalArgumentException("pattern must not be null");
        }
        this.pattern = pattern;
    }

    @Override
    public Observable<Boolean> isValid(@NonNull CharSequence text) {
        return Observable.just(pattern.matcher(text).matches());
    }
}

File: liv/src/main/java/com/benitobertoli/liv/rule/PatternRule.kt
package com.benitobertoli.liv.rule

import java.util.regex.Pattern

open class PatternRule @JvmOverloads constructor(
        override val errorMessage: CharSequence,
        private val pattern: Pattern,
        private val ignoreBlank: Boolean = true)
    : Rule {

    @JvmOverloads
    constructor(errorMessage: CharSequence,
                regex: String,
                ignoreBlank: Boolean = true)
            : this(errorMessage, Pattern.compile(regex), ignoreBlank)

    override fun isValid(input: CharSequence): Boolean {
        val matches = pattern.matcher(input).matches()
        return matches || (ignoreBlank && input.isBlank())
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PatternRule) return false

        if (errorMessage != other.errorMessage) return false
        if (pattern.pattern() != other.pattern.pattern()) return false
        if (ignoreBlank != other.ignoreBlank) return false

        return true
    }

    override fun hashCode(): Int {
        var result = errorMessage.hashCode()
        result = 31 * result + pattern.pattern().hashCode()
        result = 31 * result + ignoreBlank.hashCode()
        return result
    }


}